lib/pbench/server/unpack_tarballs.py

```diff
@@ -0,0 +1,135 @@
+        """Scans for tarballs in the TO-UNPACK subdirectories of the
+        ARCHIVE directory and unpacks them using FileTree.unpack() function.
+
+        Args:
+            lb_args: Lowerbound size of Bucket
+            ub_args: Upperbound size of Bucket
+
+        Returns:
+            Results tuple containing the counts of Total and Successful tarballs.
+        """
+        tarlist = glob.iglob(self.config.ARCHIVE / "*" / self.LINKSRC / "*.tar.xz")
+
+        tarball_list = list()
+        for tarball in tarlist:
+            tar_size = Path(tarball).stat().st_size
+            if tar_size < lb_args or tar_size > ub_args:
+                continue
+            tarball_list.append(tarball)
+
+        ntotal = nsuccess = 0
+
+        file_tree = FileTree(self.config, self.logger)
+
+        for tarball in sorted(tarball_list):
+            ntotal += 1
+
+            try:
+                tb = Path(tarball).resolve(strict=True)
+            except FileNotFoundError as exc:
+                self.logger.error(
+                    "{}: Tarball link, '{}', does not resolve to a real location: {}",
+                    self.config.TS,
+                    tarball,
+                    exc,
+                )
+                continue
+
+            try:
+                tar_md5 = get_tarball_md5(tb)
+            except FileNotFoundError as exc:
+                self.logger.error(
+                    "{}: Getting md5 value of tarball '{}' failed: {}",
+                    self.config.TS,
+                    tb,
+                    exc,
+                )
+                continue
+
+            try:
+                file_tree.unpack(tar_md5)
+            except Exception as exc:
+                self.logger.error(
+                    "{}: Unpacking of tarball {} failed: {}",
+                    self.config.TS,
+                    tb,
+                    exc,
+                )
+                continue
+
+            linkdestlist = self.config.get("pbench-server", "unpacked-states").split(
+                ","
+            )
+            controller_name = tb.parent.parent.name
+            dest = self.config.ARCHIVE / controller_name
+
+            for state in linkdestlist:
+                try:
+                    os.symlink((dest / tb.name), (dest / state / tb.name))
+                except Exception as exc:
+                    self.logger.error(
+                        "{}: Error in creation of symlink. {}", self.config.TS, exc
+                    )
+                    continue
+
+            try:
+                os.unlink(dest / self.LINKSRC / tb.name)
+            except Exception as exc:
+                self.logger.error(
+                    "{}: Error in removing symlink from {} state. {}",
+                    self.config.TS,
+                    self.LINKSRC,
+                    exc,
+                )
+                continue
+
+            nsuccess += 1
+
+        return Results(total=ntotal, success=nsuccess)
```

Comment:
This function has grown rather large; you should consider breaking it up.

As I read it, this function currently:
1. acquires a list of tarballs to unpack
2. for each tarball in that list
  a. resolve the tarball path, get the MD5, and do the unpack
  b. create the symlinks
  c. remove the source link

I would move steps 1, 2a, and 2b into separate helper functions, leaving just step 2 as the body of the original function.  (You could move 2c to a helper as well, but I don't think doing so produces much value -- it leaves the original function pretty sparse, and the helper is pretty trivial -- so I would leave it in the original function; however, there's an argument that it is related in functionality to step 2b, so you could consider moving it there.  If you move step 2c out of the original function, you could consider not moving step 1 out, but I think it will be easier to test if you do.)

In this arrangement, the step 1 helper is much easier to test, and, for testing the step 2 function you can mock the step 1 function to supply the list of tarballs.  The step 2a and 2b functions will be easier to test without the superstructure of the step 2 function above them.  And, testing the step 2 function will require fewer mocks, because you can cover all of the step 2a function with a single mock.